File: programs/practice/google_code_jam/parenthesis_sequence.py
import sys

dp = [ [0]*101 for i in range(101) ]
done = [ [0]*101 for i in range(101) ]

def f(i, j):
	if i > n or j > n or j > i:
		dp[i][j] = 0
		return 0
	if n-i == 0 or n-j == 0:
		dp[i][j] = 1
		return 1
	if done[i][j]:
		return dp[i][j]
	done[i][j] = 1
	ans = f(i+1,j) + f(i,j+1);
	dp[i][j] = ans
	return ans

tt = input()
for t in range(1,tt+1):

	sys.stdout.write('Case #%d: ' % t)
	s = map(int,raw_input().split())
	n,k = s[0],s[1]
	for i in range(101):
		for j in range(101):
			dp[i][j] = 0
			done[i][j] = 0
	f(0,0)
	dp[n][n] = 1
	
	i=0; j=0
	if k > dp[1][0]:
		print "Doesn't Exist!"
		continue
	while(k>=0):
		if k == 1:
			for a in range(n-i):
				sys.stdout.write('(')
			for a in range(n-j):
				sys.stdout.write(')')
			break

		if k > dp[i+1][j]:
			k -= dp[i+1][j]
			j+=1
			sys.stdout.write(')')
		else:
			i+=1
			sys.stdout.write('(')
	print ''
	print dp[0][0]

File: programs/practice/google_code_jam/parenthesis_sequence.cpp
// https://code.google.com/codejam/contest/4214486/dashboard#s=p3
#include <cstdio>
#include <cstring>
#include <cassert>
#include <climits>
typedef long long LL;
typedef unsigned long long ULL;

ULL dp[101][101];
char done[101][101];
int n, flag;

LL f(int i, int j)
{
	ULL &ret = dp[i][j];
	if(i > n or j > n or j > i)
		return ret = 0;
	if(n-i == 0 or n-j == 0)
		return ret = 1;
	if(done[i][j])
		return ret;
	done[i][j] = 1;
	ULL ans = f(i+1,j) + f(i,j+1);
	if(ans > 1000000000000000000LLU)
	{
		ans = 1000000000000000001LLU;
		done[i][j] = 2;
	}
	return ret = ans;
}

int main()
{
	int tt; scanf("%d",&tt);
	for(int t=1;t<=tt;t++)
	{
		printf("Case #%d: ",t);
		ULL k;
		scanf("%d %llu",&n,&k);
		memset(dp,0,sizeof(dp));
		memset(done,0,sizeof(done));
		flag = 0;
		f(0,0);
		dp[n][n] = 1;
		
		int i=0,j=0;
		if(k > dp[1][0] or flag)
		{
			printf("Doesn't Exist!\n");
			continue;
		}
		while(k>=0)
		{
			if(k == 1)
			{
				for(int a=0;a<n-i;a++)
					printf("(");
				for(int a=0;a<n-j;a++)
					printf(")");
				break;
			}

			if(k > dp[i+1][j] and done[i+1][j] != 2)
			{
				k -= dp[i+1][j];
				j++;
				printf(")");
			}
			else
			{
				i++;
				printf("(");
			}
		}
	}
	return 0;
}
